controller_stopper/src/controller_stopper.cpp

```diff
@@ -0,0 +1,163 @@
+ */
+//----------------------------------------------------------------------
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "controller_stopper/controller_stopper.hpp"
+
+ControllerStopper::ControllerStopper(const rclcpp::Node::SharedPtr& node, bool stop_controllers_on_startup)
+  : node_(node), stop_controllers_on_startup_(stop_controllers_on_startup), robot_running_(true)
+{
+  // Subscribes to a robot's running state topic. Ideally this topic is latched and only publishes
+  // on changes. However, this node only reacts on state changes, so a state published each cycle
+  // would also be fine.
+  robot_running_sub_ = node->create_subscription<std_msgs::msg::Bool>(
+      "io_and_status_controller/robot_program_running", 1,
+      std::bind(&ControllerStopper::robotRunningCallback, this, std::placeholders::_1));
+
+  // Controller manager service to switch controllers
+  controller_manager_srv_ = node_->create_client<controller_manager_msgs::srv::SwitchController>("controller_manager/"
+                                                                                                 "switch_controller");
+  // Controller manager service to list controllers
+  controller_list_srv_ = node_->create_client<controller_manager_msgs::srv::ListControllers>("controller_manager/"
+                                                                                             "list_controllers");
+
+  RCLCPP_INFO(rclcpp::get_logger("Controller stopper"), "Waiting for switch controller service to come up on "
+                                                        "controller_manager/switch_controller");
+  controller_manager_srv_->wait_for_service();
+  RCLCPP_INFO(rclcpp::get_logger("Controller stopper"), "Service available");
+  RCLCPP_INFO(rclcpp::get_logger("Controller stopper"), "Waiting for list controllers service to come up on "
+                                                        "controller_manager/list_controllers");
+  controller_list_srv_->wait_for_service();
+  RCLCPP_INFO(rclcpp::get_logger("Controller stopper"), "Service available");
+
+  consistent_controllers_ = node_->declare_parameter<std::vector<std::string>>("consistent_controllers");
+
+  if (stop_controllers_on_startup_ == true) {
+    while (stopped_controllers_.empty()) {
+      auto request = std::make_shared<controller_manager_msgs::srv::ListControllers::Request>();
+      auto future = controller_list_srv_->async_send_request(request);
+      rclcpp::spin_until_future_complete(node_, future);
+      auto result = future.get();
+      for (auto& controller : result->controller) {
+        // Check if in consistent_controllers
+        // Else:
+        //   Add to stopped_controllers
+        if (controller.state == "active") {
+          auto it = std::find(consistent_controllers_.begin(), consistent_controllers_.end(), controller.name);
+          if (it == consistent_controllers_.end() && controller.name.find("broadcaster") == std::string::npos) {
```

Comment:
I would rather not use such a string matching but rather rely on the consistent_controllers list solely.
```suggestion
          if (it == consistent_controllers_.end()) {
```